src/retro_data_structures/disc/game_disc.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+
+import collections
+import dataclasses
+import io
+import typing
+
+import construct
+
+from retro_data_structures.disc import disc_common
+from retro_data_structures.disc.gc_disc import GcDisc
+from retro_data_structures.disc.wii_disc import WiiDisc, WiiPartition
+from retro_data_structures.formats import dol
+
+if typing.TYPE_CHECKING:
+    from pathlib import Path
+
+
+class UnsupportedDiscFormat(Exception):
+    pass
+
+
+@dataclasses.dataclass
+class FileEntry:
+    offset: int
+    size: int
+
+
+FileTree: typing.TypeAlias = dict[str, typing.Union[FileEntry, "FileTree"]]
+
+
+def decode_into_file_tree(fst) -> FileTree:
+    file_tree: dict = {}
+    current_dir = file_tree
+
+    end_folder = collections.defaultdict(list)
+
+    names_stream = io.BytesIO(fst.names)
+    for i, file in enumerate(fst.file_entries):
+        if i == 0:
+            continue
+
+        if i in end_folder:
+            current_dir = end_folder.pop(i)[0]
+
+        names_stream.seek(file.file_name)
+        name = construct.CString("ascii").parse_stream(names_stream)
+        if file.is_directory:
+            new_dir = {}
+            end_folder[file.param].append(current_dir)
+            current_dir[name] = new_dir
+            current_dir = new_dir
+        else:
+            current_dir[name] = FileEntry(
+                offset=file.offset,
+                size=file.param,
+            )
+
+    return file_tree
+
+
+CommonGCWiiHeader = construct.Struct(
+    game_code=construct.Bytes(4),
+    maker_code=construct.Bytes(2),
+    disc_id=construct.Int8ub,  # for multi-disc games
+    version=construct.Int8ub,
+    audio_streaming=construct.Int8ub,
+    stream_buffer_size=construct.Int8ub,
+    _unused_a=construct.Const(b"\x00" * 14),
+    wii_magic_word=construct.Int32ub,  # 0x5D1C9EA3
+    gc_magic_word=construct.Int32ub,  # 0xC2339F3D
+)
```

Comment:
any reason to parse the whole common header if we only need the two magic words? could just as easily use a `Peek()` or something, no?